hamilton/experimental/h_async.py

```diff
@@ -57,34 +77,111 @@ def execute_node(self, node: node.Node, kwargs: typing.Dict[str, typing.Any]) ->
         Note that this assumes that everything is awaitable, even if it isn't.
         In that case, it just wraps it in one.
 
+        :param task_id:
+        :param node_:
+        :param run_id:
         :param node: Node to wrap
         :param kwargs: Keyword arguments (either coroutines or raw values) to call it with
         :return: A task
         """
-        callabl = node.callable
+        callabl = node_.callable
 
         async def new_fn(fn=callabl, **fn_kwargs):
             task_dict = {key: process_value(value) for key, value in fn_kwargs.items()}
             fn_kwargs = await await_dict_of_tasks(task_dict)
-            if inspect.iscoroutinefunction(fn):
-                return await fn(**fn_kwargs)
-            return fn(**fn_kwargs)
+            error = None
+            result = None
+            success = True
+            pre_node_execute_errored = False
+            try:
+                if self.adapter.does_hook("pre_node_execute", is_async=True):
+                    try:
+                        await self.adapter.call_all_lifecycle_hooks_async(
+                            "pre_node_execute",
+                            run_id=run_id,
+                            node_=node_,
+                            kwargs=fn_kwargs,
+                            task_id=task_id,
+                        )
+                    except Exception as e:
+                        pre_node_execute_errored = True
+                        raise e
+                # TODO -- consider how to use node execution methods in the future
+                # This is messy as it is a method called within a method...
+                # if self.adapter.does_method("do_node_execute", is_async=False):
+                #     result = self.adapter.call_lifecycle_method_sync(
+                #         "do_node_execute",
+                #         run_id=run_id,
+                #         node_=node_,
+                #         kwargs=kwargs,
+                #         task_id=task_id,
+                #     )
+                # else:
```

Comment:
?

Reply:
Methods versus hooks. Currently we don't respect methods (E.G. `do_node_execute`), as we're using a hook to implement this.